File: trade_on_dex.js
const xrpl = require("xrpl");
const fs = require("fs");

const TOKEN_CODE = "TES";
const SELL_AMOUNT = "2"; 
const BUY_AMOUNT = "2"; 
const PRICE_PER_TOKEN_XRP = "1"; 

async function main() {
  const client = new xrpl.Client("wss://s.altnet.rippletest.net:51233");
  await client.connect();

  // Load wallets
  const wallets = JSON.parse(fs.readFileSync("wallets.json"));
  const issuer = xrpl.Wallet.fromSeed(wallets.find(w => w.name === "issuer").seed);
  const receiver = xrpl.Wallet.fromSeed(wallets.find(w => w.name === "receiver").seed);
  const third = xrpl.Wallet.fromSeed(wallets.find(w => w.name === "third").seed);

  // 1. Receiver places a sell offer (sell 50 TES for XRP)
  const sellOfferTx = {
    TransactionType: "OfferCreate",
    Account: receiver.classicAddress,
    TakerGets: {
      currency: TOKEN_CODE,
      issuer: issuer.classicAddress,
      value: SELL_AMOUNT,
    },
    TakerPays: xrpl.xrpToDrops((parseFloat(SELL_AMOUNT) * parseFloat(PRICE_PER_TOKEN_XRP)).toString()),
  };
  const sell_prepared = await client.autofill(sellOfferTx);
  const sell_signed = receiver.sign(sell_prepared);
  const sell_result = await client.submitAndWait(sell_signed.tx_blob);
  console.log("Sell offer placed by receiver:", sell_result.result.meta.TransactionResult);

  // 2. Third places a buy offer (buy 50 TES for XRP)
  const buyOfferTx = {
    TransactionType: "OfferCreate",
    Account: third.classicAddress,
    TakerGets: xrpl.xrpToDrops((parseFloat(BUY_AMOUNT) * parseFloat(PRICE_PER_TOKEN_XRP)).toString()),
    TakerPays: {
      currency: TOKEN_CODE,
      issuer: issuer.classicAddress,
      value: BUY_AMOUNT,
    },
  };
  const buy_prepared = await client.autofill(buyOfferTx);
  const buy_signed = third.sign(buy_prepared);
  const buy_result = await client.submitAndWait(buy_signed.tx_blob);
  console.log("Buy offer placed by third:", buy_result.result.meta.TransactionResult);

  await client.disconnect();
}

main().catch(console.error);

File: trade_on_dex.py
import asyncio
import json
from decimal import Decimal
from pathlib import Path
from typing import Dict, List

from xrpl.models.transactions import OfferCreate
from xrpl.utils import xrp_to_drops
from xrpl.wallet import Wallet

from xrpl_helpers import issued_currency_amount, submit_and_log, xrpl_client

TOKEN_CODE = "TES"
SELL_AMOUNT = Decimal("2")
BUY_AMOUNT = Decimal("2")
PRICE_PER_TOKEN_XRP = Decimal("1")
WALLETS_PATH = Path("wallets.json")


def load_wallet(name: str, records: List[Dict[str, str]]) -> Wallet:
    record = next((entry for entry in records if entry["name"] == name), None)
    if record is None:
        raise ValueError(f"Wallet with name '{name}' not found in {WALLETS_PATH}")
    return Wallet.from_seed(record["seed"])


async def main() -> None:
    if not WALLETS_PATH.exists():
        raise FileNotFoundError(
            f"Missing {WALLETS_PATH}. Run issue_token.py to generate funded wallets."
        )

    wallets = json.loads(WALLETS_PATH.read_text())
    issuer = load_wallet("issuer", wallets)
    receiver = load_wallet("receiver", wallets)
    third = load_wallet("third", wallets)
    xrp_total = SELL_AMOUNT * PRICE_PER_TOKEN_XRP

    async with xrpl_client() as client:
        await submit_and_log(
            client,
            OfferCreate(
                account=receiver.classic_address,
                taker_gets=issued_currency_amount(
                    TOKEN_CODE, issuer.classic_address, str(SELL_AMOUNT)
                ),
                taker_pays=xrp_to_drops(str(xrp_total)),
            ),
            receiver,
            description="Receiver placed a sell offer for TES",
        )

        await submit_and_log(
            client,
            OfferCreate(
                account=third.classic_address,
                taker_gets=xrp_to_drops(str(xrp_total)),
                taker_pays=issued_currency_amount(
                    TOKEN_CODE, issuer.classic_address, str(BUY_AMOUNT)
                ),
            ),
            third,
            description="Third placed a buy offer for TES",
        )


if __name__ == "__main__":
    asyncio.run(main())
